### PR TITLE
Refactor QueryBugsTest to use a non-provider specific way to configure the contexts

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Part of #17588

### CODE REVIEW

File: test/EFCore.Specification.Tests/NonSharedModelTestBase.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        private TestStore _testStore;
+        protected TestStore TestStore
+            => _testStore ?? throw new InvalidOperationException(
+                $"You must call `await {nameof(InitializeAsync)}(\"DatabaseName\");` at the beggining of the test.");
+
+        private ListLoggerFactory _listLoggerFactory;
+        protected ListLoggerFactory ListLoggerFactory
+            => _listLoggerFactory ??= (ListLoggerFactory)ServiceProvider.GetRequiredService<ILoggerFactory>();
+
+        public virtual Task InitializeAsync() => Task.CompletedTask;
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/NonSharedModelTestBase.cs:35]: Where is this used?

File: test/EFCore.Specification.Tests/NonSharedModelTestBase.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            Action<ModelBuilder> onModelCreating = null,
+            Action<DbContextOptionsBuilder> onConfiguring = null,
+            Action<IServiceCollection> addServices = null,
+            Action<TContext> seed = null,
+            Func<string, bool> shouldLogCategory = null,
+            Func<TestStore> createTestStore = null,
+            bool usePooling = true)
+            where TContext : DbContext
+        {
+            var contextFactory = Initialize<TContext>(
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/NonSharedModelTestBase.cs:67]: If this body is not doing anything async, is there any reason for us to use this particular method everywhere and make tests async (even though all tests are running queries in sync)?

Comment (AndriySvyryd) [reply test/EFCore.Specification.Tests/NonSharedModelTestBase.cs:67]: It's in preparation for adding `TestStore.InitializeAsync`

File: test/EFCore.Specification.Tests/NonSharedModelTestBase.cs (hunk 1 above)
Comment (AndriySvyryd) [reply test/EFCore.Specification.Tests/NonSharedModelTestBase.cs:35]: Part of `IAsyncLifetime`

**Review by smitpatel — APPROVED** [review]

File: test/EFCore.Specification.Tests/NonSharedModelTestBase.cs (hunk 2 above)
Comment (smitpatel) [reply test/EFCore.Specification.Tests/NonSharedModelTestBase.cs:67]: Does that mean tests can call either InitializeAsync or Initialize?

Comment (AndriySvyryd) [reply test/EFCore.Specification.Tests/NonSharedModelTestBase.cs:67]: Yes

**Pull request merged**